pkg/virt-operator/kubevirt.go

```diff
@@ -422,6 +456,211 @@ func (c *KubeVirtController) execute(key string) error {
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c.installStrategyMutex.Lock()
+	defer c.installStrategyMutex.Unlock()
+	// reset the local map
+	c.installStrategyMap = make(map[string]*installstrategy.InstallStrategy)
+
+	return nil
+}
+
+// Loads install strategies into memory, and generates jobs to
+// create install strategies that don't exist yet.
+func (c *KubeVirtController) loadInstallStrategy(kv *v1.KubeVirt) (*installstrategy.InstallStrategy, bool, error) {
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// 1. see if we already loaded the install strategy
+	strategy, ok := c.getInstallStrategyFromMap(c.config.ImageTag)
+	if ok {
+		// we already loaded this strategy into memory
+		return strategy, false, nil
+	}
+
+	// 2. look for install strategy config map in cache.
+	strategy, err = installstrategy.LoadInstallStrategyFromCache(c.stores, kv.Namespace, c.config.ImageTag)
+	if err == nil {
+		c.installStrategyMutex.Lock()
+		defer c.installStrategyMutex.Unlock()
+		c.installStrategyMap[c.config.ImageTag] = strategy
+		log.Log.Infof("Loaded install strategy for kubevirt version %s into cache", c.config.ImageTag)
+		return strategy, false, nil
+	}
+
+	log.Log.Infof("Install strategy config map not loaded. reason: %v", err)
+
+	// 3. See if we have a pending job in flight for this install strategy.
+	batch := c.clientset.BatchV1()
+	job := c.generateInstallStrategyJob(kv, c.config.ImageTag, c.config.ImageRegistry)
+
+	obj, exists, _ := c.stores.InstallStrategyJobCache.Get(job)
+	if exists {
+		cachedJob := obj.(*batchv1.Job)
+
+		if cachedJob.Status.CompletionTime != nil {
+			// job completed but we don't have a install strategy still
+			// delete the job and we'll re-execute it once it is removed.
+
+			log.Log.Object(cachedJob).Errorf("Job failed to create install strategy for version %s", c.config.ImageTag)
+			if cachedJob.DeletionTimestamp == nil {
+
+				// Just in case there's an issue causing the job to fail
+				// immediately after being posted, lets perform a rudimentary
+				// for of rate-limiting for how quickly we'll re-attempt.
+				// TODO there's an alpha feature that lets us set a TTL on the job
+				// itself which will ensure it is automatically cleaned up for us
+				// after completion. That feature is feature-gated and isn't something
+				// we can depend on right now though.
+				now := time.Now().UTC().Unix()
+				secondsSinceCompletion := now - cachedJob.Status.CompletionTime.UTC().Unix()
+				if secondsSinceCompletion < 10 {
+					secondsLeft := int64(10)
+					if secondsSinceCompletion > 0 {
+						secondsLeft = secondsSinceCompletion
+					}
+					c.queue.AddAfter(kvkey, time.Duration(secondsLeft)*time.Second)
+
+				} else {
+					key, err := controller.KeyFunc(cachedJob)
+					if err != nil {
+						return nil, true, err
+					}
+
+					c.kubeVirtExpectations.InstallStrategyJob.AddExpectedDeletion(kvkey, key)
```

Comment:
As it is right now, I don't think that you would need an expecter for the job. We create it, it or the configmap is the only one which would wake us up again. Only then we continue and start creating other objects. That's slightly different than for the rest, where we create a lot of objects without returning from the controller. If we encounter issues there, we only see parts of the updates. Idon't mind if the expecter is there.

Reply:
> As it is right now, I don't think that you would need an expecter for the job

if we don't have the expecter, we could create multiple active jobs. if the queue is executed again before we observe the job creation. that's not a huge deal, but it is more efficient to have the expecter here. 

we need the job to wake us up again in case the job fails.